src/tls12-download.c

```diff
@@ -326,10 +329,58 @@ int __stdcall entry()
     WinHttpCloseHandle(request);
     WinHttpCloseHandle(connect);
     WinHttpCloseHandle(session);
-    CloseHandle(out_file);
 
-    write_message(std_out, L"Done.\r\n");
+    write_message(std_out, L" done.\r\nValidating signature...");
+
+    if (SetFilePointer(out_file, 0, 0, FILE_BEGIN) == INVALID_SET_FILE_POINTER)
+    {
+        abort_api_failure(std_out, L"SetFilePointer");
+    }
+
+    DWORD exit_code = 0;
+
+    WINTRUST_FILE_INFO wtfi;
+    wtfi.cbStruct = sizeof(wtfi);
+    wtfi.pcwszFilePath = out_file_path;
+    wtfi.hFile = out_file;
+    wtfi.pgKnownSubject = 0;
+
+    WINTRUST_DATA wtd;
+    ZeroMemory(&wtd, sizeof(wtd));
+    wtd.cbStruct = sizeof(wtd);
+    wtd.dwUIChoice = WTD_UI_NONE;
+    wtd.fdwRevocationChecks = WTD_REVOKE_WHOLECHAIN;
+    wtd.dwUnionChoice = WTD_CHOICE_FILE;
+    wtd.pFile = &wtfi;
+    wtd.dwStateAction = WTD_STATEACTION_VERIFY;
+    wtd.dwProvFlags = WTD_REVOCATION_CHECK_CHAIN;
+
+    GUID wt_policy_guid = WINTRUST_ACTION_GENERIC_VERIFY_V2;
+
+    LONG trust_validation_result = WinVerifyTrust(INVALID_HANDLE_VALUE, &wt_policy_guid, &wtd);
+    wtd.dwStateAction = WTD_STATEACTION_CLOSE;
+    (void)WinVerifyTrust(INVALID_HANDLE_VALUE, &wt_policy_guid, &wtd);
+    if (trust_validation_result == 0)
+    {
+        write_message(std_out, L" done.\r\n");
+    }
+    else
+    {
+        exit_code = 1;
+        write_message(std_out, L" failed! ");
+        write_hex(std_out, (DWORD)trust_validation_result);
+        FILE_DISPOSITION_INFO fdi;
+        fdi.DeleteFile = TRUE;
+        if (SetFileInformationByHandle(out_file, FileDispositionInfo, &fdi, sizeof(fdi)) == 0)
```

Comment:
Would it make sense to pend the delete before downloading and only un-pend the delete once the signature has been validated?

Reply:
Done